src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_GetLatestDiagnosticsForSpan.cs

```diff
@@ -107,18 +109,137 @@ private class LatestDiagnosticsForSpanGetter
+                    (t, d) => t.Equals(textVersion) && d.Equals(semanticVersion),
+                    async (_1, _2) =>
+                    {
+                        var model = await _document.GetSemanticModelAsync(_cancellationToken).ConfigureAwait(false);
+                        VerifyDiagnostics(model);
+
+                        var root = await _document.GetSyntaxRootAsync(_cancellationToken).ConfigureAwait(false);
+                        var adjustedSpan = AdjustSpan(_document, root, _range);
+                        var diagnostics = model.GetDeclarationDiagnostics(adjustedSpan, _cancellationToken).Concat(model.GetMethodBodyDiagnostics(adjustedSpan, _cancellationToken));
+
+                        return GetDiagnosticData(_document, model.SyntaxTree, _range, diagnostics);
+                    }).ConfigureAwait(false);
+
+                return containsFullResult;
+            }
+
+            [Conditional("DEBUG")]
+            private void VerifyDiagnostics(SemanticModel model)
+            {
+                Func<Diagnostic, bool> shouldInclude = d => _range.IntersectsWith(d.Location.SourceSpan);
+
+                // make sure what we got from range is same as what we got from whole diagnostics
+                var rangeDeclaractionDiagnostics = model.GetDeclarationDiagnostics(_range).ToArray();
+                var rangeMethodBodyDiagnostics = model.GetMethodBodyDiagnostics(_range).ToArray();
+                var rangeDiagnostics = rangeDeclaractionDiagnostics.Concat(rangeMethodBodyDiagnostics).Where(shouldInclude).ToArray();
+
+                var set = new HashSet<Diagnostic>(rangeDiagnostics);
+
+                var wholeDeclarationDiagnostics = model.GetDeclarationDiagnostics().ToArray();
+                var wholeMethodBodyDiagnostics = model.GetMethodBodyDiagnostics().ToArray();
+                var wholeDiagnostics = wholeDeclarationDiagnostics.Concat(wholeMethodBodyDiagnostics).Where(shouldInclude).ToArray();
+
+                if (!set.SetEquals(wholeDiagnostics))
+                {
+                    // otherwise, report non-fatal watson so that we can fix those cases
+                    FatalError.ReportWithoutCrash(new Exception("Bug in GetDiagnostics"));
+
+                    // make sure we hold onto these even in ret build for debugging.
+                    GC.KeepAlive(set);
+                    GC.KeepAlive(rangeDeclaractionDiagnostics);
+                    GC.KeepAlive(rangeMethodBodyDiagnostics);
+                    GC.KeepAlive(rangeDiagnostics);
+                    GC.KeepAlive(wholeDeclarationDiagnostics);
+                    GC.KeepAlive(wholeMethodBodyDiagnostics);
+                    GC.KeepAlive(wholeDiagnostics);
+                }
+            }
+
+            private static TextSpan AdjustSpan(Document document, SyntaxNode root, TextSpan span)
+            {
+                // this is to workaround a bug (https://github.com/dotnet/roslyn/issues/1557)
```

Comment:
Did you find any specific repro case for this? If so, can you please add it to #1557?